Intermittent error while attempting to sort code coverage beans

I'm occasionally experiencing errors at the point at which code coverage beans are sorted in ascending order of coverage percentage, prior to being written to the report file. Here's an example of the error:
```
[com.sforce.cd.apexUnit.ApexUnitRunner.main()] INFO com.sforce.cd.apexUnit.ApexUnitRunner - ####################################   Processing the Apex test classes specified by the user  #################################### 
[com.sforce.cd.apexUnit.ApexUnitRunner.main()] INFO com.sforce.cd.apexUnit.client.utils.ApexClassFetcherUtils - Using regex(es): Test_Service_Fee to fetch apex classes
[com.sforce.cd.apexUnit.ApexUnitRunner.main()] INFO com.sforce.cd.apexUnit.client.utils.ApexClassFetcherUtils - Using regex: "Test_Service_Fee" to fetch apex classes
[com.sforce.cd.apexUnit.ApexUnitRunner.main()] INFO com.sforce.cd.apexUnit.client.connection.ConnectionHandler - Bulk connection established.
[com.sforce.cd.apexUnit.ApexUnitRunner.main()] INFO com.sforce.cd.apexUnit.client.testEngine.AsyncBulkApiHandler - Creating batch for the test classes to execute using bulk connection....
[com.sforce.cd.apexUnit.ApexUnitRunner.main()] INFO com.sforce.cd.apexUnit.client.testEngine.AsyncBulkApiHandler - Async test parent job Id : 707d0000035JkJuAAK
[com.sforce.cd.apexUnit.ApexUnitRunner.main()] INFO com.sforce.cd.apexUnit.client.testEngine.AsyncBulkApiHandler - ############################# List of Apex test classes successfully submitted to the Force.com test execution engine #############################
[com.sforce.cd.apexUnit.ApexUnitRunner.main()] INFO com.sforce.cd.apexUnit.client.testEngine.TestExecutor - Parent job ID for the submission of the test classes to the Force.com platform is: 707d0000035JkJuAAK
[com.sforce.cd.apexUnit.ApexUnitRunner.main()] INFO com.sforce.cd.apexUnit.client.testEngine.TestExecutor - ############################# Now executing - Apex tests.. #############################
[com.sforce.cd.apexUnit.ApexUnitRunner.main()] INFO com.sforce.cd.apexUnit.client.testEngine.TestStatusPollerAndResultHandler - Total test classes to execute: 1
[com.sforce.cd.apexUnit.ApexUnitRunner.main()] INFO com.sforce.cd.apexUnit.client.testEngine.TestStatusPollerAndResultHandler - Now executing the test class: Test_Service_Fee (https://na14.salesforce.com/01pd00000032yyKAAQ ) Status : Processing
[com.sforce.cd.apexUnit.ApexUnitRunner.main()] INFO com.sforce.cd.apexUnit.client.testEngine.TestStatusPollerAndResultHandler - Test status for Test_Service_Fee:(24/24)
[com.sforce.cd.apexUnit.ApexUnitRunner.main()] INFO com.sforce.cd.apexUnit.client.testEngine.TestStatusPollerAndResultHandler - Completed executing the test class: Test_Service_Fee. Time taken by the test: 0 minutes,26 seconds
[com.sforce.cd.apexUnit.ApexUnitRunner.main()] INFO com.sforce.cd.apexUnit.client.testEngine.TestStatusPollerAndResultHandler - Total tests executed 1 , Remaining tests 0
[com.sforce.cd.apexUnit.ApexUnitRunner.main()] INFO com.sforce.cd.apexUnit.client.testEngine.TestStatusPollerAndResultHandler - All tests have now completed executing!!
[com.sforce.cd.apexUnit.ApexUnitRunner.main()] INFO com.sforce.cd.apexUnit.ApexUnitRunner - ####################################    Computing code coverage for the team based on the Apex Class names(source class names) provided  #################################### 
[com.sforce.cd.apexUnit.ApexUnitRunner.main()] INFO com.sforce.cd.apexUnit.client.fileReader.ApexManifestFileReader - Reading from Manifest files: patronticket_classes.txt
[com.sforce.cd.apexUnit.ApexUnitRunner.main()] INFO com.sforce.cd.apexUnit.client.fileReader.ApexManifestFileReader - Reading Manifest file from location : patronticket_classes.txt
Jan 08, 2016 9:03:28 AM org.apache.commons.httpclient.HttpMethodBase getResponseBody
WARNING: Going to buffer response body of large or unknown size. Using getResponseBodyAsStream instead is recommended.
[com.sforce.cd.apexUnit.ApexUnitRunner.main()] INFO com.sforce.cd.apexUnit.client.codeCoverage.WebServiceInvoker - Status code : 200   Status message from the get request:OK Reason phrase: OK
Jan 08, 2016 9:03:29 AM org.apache.commons.httpclient.HttpMethodBase getResponseBody
WARNING: Going to buffer response body of large or unknown size. Using getResponseBodyAsStream instead is recommended.
[com.sforce.cd.apexUnit.ApexUnitRunner.main()] INFO com.sforce.cd.apexUnit.client.codeCoverage.CodeCoverageComputer - Record number # 1 : coveredLines : 6 : unCoveredLines : 19 : code coverage % : 24.0 : apexClassOrTriggerId : 01pd0000002roPpAAI : apexClassName : MSU_SectionGroupWrapper : apiVersion : 34.0 : lengthWithoutComments : 1323
[com.sforce.cd.apexUnit.ApexUnitRunner.main()] INFO com.sforce.cd.apexUnit.client.codeCoverage.CodeCoverageComputer - Record number # 2 : coveredLines : 87 : unCoveredLines : 14 : code coverage % : 86.13861386138613 : apexClassOrTriggerId : 01pd0000002roPkAAI : apexClassName : MSU_SectionWrapper : apiVersion : 34.0 : lengthWithoutComments : 5662
[com.sforce.cd.apexUnit.ApexUnitRunner.main()] INFO com.sforce.cd.apexUnit.client.codeCoverage.CodeCoverageComputer - Record number # 3 : coveredLines : 32 : unCoveredLines : 10 : code coverage % : 76.19047619047619 : apexClassOrTriggerId : 01pd0000002r0cNAAQ : apexClassName : TicketAllocationsSelector : apiVersion : 29.0 : lengthWithoutComments : 3357
.
.
.
[com.sforce.cd.apexUnit.ApexUnitRunner.main()] INFO com.sforce.cd.apexUnit.client.codeCoverage.CodeCoverageComputer - Record number # 200 : coveredLines : 292 : unCoveredLines : 78 : code coverage % : 78.91891891891892 : apexClassOrTriggerId : 01pd0000002r0cOAAQ : apexClassName : TicketOrderItems : apiVersion : 29.0 : lengthWithoutComments : 26323
[com.sforce.cd.apexUnit.ApexUnitRunner.main()] INFO com.sforce.cd.apexUnit.client.codeCoverage.CodeCoverageComputer - Record number # 201 : coveredLines : 1953 : unCoveredLines : 108 : code coverage % : 94.75982532751091 : apexClassOrTriggerId : 01pd0000002r0cHAAQ : apexClassName : Service_Transaction : apiVersion : 29.0 : lengthWithoutComments : 135166
[com.sforce.cd.apexUnit.ApexUnitRunner.main()] INFO com.sforce.cd.apexUnit.client.codeCoverage.CodeCoverageComputer - Record number # 202 : coveredLines : 52 : unCoveredLines : 11 : code coverage % : 82.53968253968253 : apexClassOrTriggerId : 01pd0000002r0cIAAQ : apexClassName : SubscriptionBuyerSelectionLinksSelector : apiVersion : 29.0 : lengthWithoutComments : 4799
[com.sforce.cd.apexUnit.ApexUnitRunner.main()] INFO com.sforce.cd.apexUnit.client.codeCoverage.CodeCoverageComputer - Record number # 203 : coveredLines : 192 : unCoveredLines : 1 : code coverage % : 99.48186528497409 : apexClassOrTriggerId : 01pd0000002r0cGAAQ : apexClassName : Service_Payment : apiVersion : 29.0 : lengthWithoutComments : 13124
[com.sforce.cd.apexUnit.ApexUnitRunner.main()] INFO com.sforce.cd.apexUnit.client.codeCoverage.CodeCoverageComputer - Record number # 204 : coveredLines : 180 : unCoveredLines : 11 : code coverage % : 94.24083769633508 : apexClassOrTriggerId : 01pd0000002r0cPAAQ : apexClassName : TicketOrderItemsSelector : apiVersion : 29.0 : lengthWithoutComments : 14460
[com.sforce.cd.apexUnit.ApexUnitRunner.main()] INFO com.sforce.cd.apexUnit.client.codeCoverage.CodeCoverageComputer - Record number # 205 : coveredLines : 207 : unCoveredLines : 20 : code coverage % : 91.18942731277532 : apexClassOrTriggerId : 01pd0000002r0caAAA : apexClassName : fflib_SObjectDomain : apiVersion : 31.0 : lengthWithoutComments : 13240
[com.sforce.cd.apexUnit.ApexUnitRunner.main()] INFO com.sforce.cd.apexUnit.client.codeCoverage.CodeCoverageComputer - ####################################   Summary of code coverage computation for the team..  #################################### 
[com.sforce.cd.apexUnit.ApexUnitRunner.main()] INFO com.sforce.cd.apexUnit.client.codeCoverage.CodeCoverageComputer - Total Covered lines : 31282
 Total Uncovered lines : 8963
[com.sforce.cd.apexUnit.ApexUnitRunner.main()] INFO com.sforce.cd.apexUnit.client.codeCoverage.CodeCoverageComputer - Team code coverage is : 77.72891042365512%
[WARNING] 
java.lang.reflect.InvocationTargetException
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:497)
	at org.codehaus.mojo.exec.ExecJavaMojo$1.run(ExecJavaMojo.java:297)
	at java.lang.Thread.run(Thread.java:745)
Caused by: java.lang.IllegalArgumentException: Comparison method violates its general contract!
	at java.util.ComparableTimSort.mergeLo(ComparableTimSort.java:744)
	at java.util.ComparableTimSort.mergeAt(ComparableTimSort.java:481)
	at java.util.ComparableTimSort.mergeForceCollapse(ComparableTimSort.java:422)
	at java.util.ComparableTimSort.sort(ComparableTimSort.java:222)
	at java.util.Arrays.sort(Arrays.java:1246)
	at com.sforce.cd.apexUnit.ApexUnitRunner.main(ApexUnitRunner.java:111)
	... 6 more
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time: 2:24.903s
[INFO] Finished at: Fri Jan 08 09:04:27 EST 2016
[INFO] Final Memory: 10M/149M
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal org.codehaus.mojo:exec-maven-plugin:1.2.1:java (default-cli) on project ApexUnit-core: An exception occured while executing the Java class. null: InvocationTargetException: Comparison method violates its general contract! -> [Help 1]
[ERROR] 
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR] 
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoExecutionException
```

I believe this error is due to the fact that the ApexClassCodeCoverageBean.compareTo method is implemented as follows:
```	
public int compareTo(ApexClassCodeCoverageBean codeCoverageBean) {
		// sort in ascending order
		return (int) (this.getCoveragePercentage() - codeCoverageBean.getCoveragePercentage());
}
```

Here is an excerpt from the Javadocs for the compareTo method:
```
Compares this object with the specified object for order. Returns a negative integer, zero, or a positive integer as this object is less than, equal to, or greater than the specified object.
The implementor must ensure sgn(x.compareTo(y)) == -sgn(y.compareTo(x)) for all x and y. (This implies that x.compareTo(y) must throw an exception iff y.compareTo(x) throws an exception.)

The implementor must also ensure that the relation is transitive: (x.compareTo(y)>0 && y.compareTo(z)>0) implies x.compareTo(z)>0.

Finally, the implementor must ensure that x.compareTo(y)==0 implies that sgn(x.compareTo(z)) == sgn(y.compareTo(z)), for all z.

It is strongly recommended, but not strictly required that (x.compareTo(y)==0) == (x.equals(y)). Generally speaking, any class that implements the Comparable interface and violates this condition should clearly indicate this fact. The recommended language is "Note: this class has a natural ordering that is inconsistent with equals."

In the foregoing description, the notation sgn(expression) designates the mathematical signum function, which is defined to return one of -1, 0, or 1 according to whether the value of expression is negative, zero or positive.

```

I suspect this method of comparing object is not adhering to the transitive contract, since there is some integer truncation error in this method. Also, consider "this" bean with a code coverage of 50.1%, and the "other" bean with a code coverage of 50.2%. With this implementation, the compareTo method would return 0 (equals) even though they are clearly not. I suggest changing this method to the following:

```
public int compareTo(ApexClassCodeCoverageBean codeCoverageBean) {
		if (this.getCoveragePercentage() < codeCoverageBean.getCoveragePercentage()) return -1;
		if (this.getCoveragePercentage() > codeCoverageBean.getCoveragePercentage()) return 1;
		return 0;
}
```

--- a/src/main/java/com/sforce/cd/apexUnit/report/ApexClassCodeCoverageBean.java
+++ b/src/main/java/com/sforce/cd/apexUnit/report/ApexClassCodeCoverageBean.java
@@ -145,7 +145,8 @@ public class ApexClassCodeCoverageBean implements Comparable<ApexClassCodeCovera
 	 * @see java.lang.Comparable#compareTo(java.lang.Object)
 	 */
 	public int compareTo(ApexClassCodeCoverageBean codeCoverageBean) {
-		// sort in ascending order
-		return (int) (this.getCoveragePercentage() - codeCoverageBean.getCoveragePercentage());
+		if (this.getCoveragePercentage() < codeCoverageBean.getCoveragePercentage()) return -1;
+		if (this.getCoveragePercentage() > codeCoverageBean.getCoveragePercentage()) return 1;
+		return 0;
 	}
 }

